### PR TITLE
storage: improve index lookups

### PR DESCRIPTION
**Pull request by beorn7**
@fabxc @juliusv Here is what I teased you with during lunch.

tl;dr: This is not a fundamental solution to the indexing problem
(like tindex is) but it at least avoids utilizing the intersection
problem to the greatest possible amount.

In more detail:

Imagine the following query:

    nicely:aggregating:rule{job="foo",env="prod"}

While it uses a nicely aggregating recording rule (which might have a
very low cardinality), Prometheus still intersects the low number of
fingerprints for `{__name__="nicely:aggregating:rule"}` with the many
thousands of fingerprints matching `{job="foo"}` and with the millions
of fingerprints matching `{env="prod"}`. This totally innocuous query
is dead slow if the Prometheus server has a lot of time series with
the `{env="prod"}` label. Ironically, if you make the query more
complicated, it becomes blazingly fast:

    nicely:aggregating:rule{job=~"foo",env=~"prod"}

Why so? Because Prometheus only intersects with non-Equal matchers if
there are no Equal matchers. That's good in this case because it
retrieves the few fingerprints for
`{__name__="nicely:aggregating:rule"}` and then starts right ahead to
retrieve the metric for those FPs and checking individually if they
match the other matchers.

This change is generalizing the idea of when to stop intersecting FPs
and go into "retrieve metrics and check them individually against
remaining matchers" mode:

- First, sort all matchers by "expected cardinality". Matchers
  matching the empty string are always worst (and never used for
  intersections). Equal matchers are in general consider best, but by
  using some crude heuristics, we declare some better than others
  (instance labels or anything that looks like a recording rule).

- Then go through the matchers until we hit a threshold of remaining
  FPs in the intersection. This threshold is higher if we are already
  in the non-Equal matcher area as intersection is even more expensive
  here.

- Once the threshold has been reached (or we have run out of matchers
  that do not match the empty string), start with "retrieve metrics
  and check them individually against remaining matchers".

A beefy server at SoundCloud was spending 67% of its CPU time in index
lookups (fingerprintsForLabelPairs), serving mostly a dashboard that
is exclusively built with recording rules. With this change, it spends
only 35% in fingerprintsForLabelPairs. The CPU usage dropped from 26
cores to 18 cores. The median latency for query_range dropped from 14s
to 50ms(!). As expected, higher percentile latency didn't improve that
much because the new approach is _occasionally_ running into the worst
case while the old one was _systematically_ doing so. The 99th
percentile latency is now about as high as the median before (14s)
while it was almost twice as high before (26s).

### CODE REVIEW

**Comment by beorn7** [comment]
prometheus-stream

**Comment by matthiasr** [comment]
What about the `__name__` label? I'd expect this to always have a pretty low cardinality.

**Comment by beorn7** [comment]
I was thinking about `__name__` but in the end, it's not much of a signal. There are many metric names that are used a lot across systems (e.g. http_request_duration), and in particular you get new time series of the same name in the typical "series churn by deployment" scenario.

**Comment by beorn7** [comment]
I'd like to try this out at larger scale at SC and tweak it a bit. So it would be great to get an opinion if this is sufficiently sane to pursue.

**Comment by brian-brazil** [comment]
This in the first version is already getting fairly magic, I think it'd be saner maintenance-wise to keep it simple - e.g. do a probe for number of matches for each equality matcher, pick the smallest and then apply the rest of the matchers.

**Comment by beorn7** [comment]
@brian-brazil That was my thinking, too, but as it looks, leveldb has no cheap way of counting the number of returned items, i.e. counting would be the same cost as retrieving. We would only save on the intersection, but from all I can see, it's the retrieval that kills us here.

**Comment by beorn7** [comment]
To expand a bit: This is a two-step improvement:
1.The "scientific" one: Stop intersecting on equal matchers once we have narrowed down the choice to 1000 FPs, and in turn apply regexp/unequal matchers if the equal matchers couldn't narrow down to fewer than 10,000 FPs.
2. The "magic" one: Instead of random order of matchers to apply, use some heuristics.

We could of course just use (1) and disregard (2) because of "magic". On the other hand, there is little to lose, and the "magic" is well contained on one method (calculateScore).

**Comment by brian-brazil** [comment]
I think a large warning in `calculateScore` that this is an interim solution, and tweaks won't be accepted would handle most of it.

**Comment by beorn7** [comment]
Updated the doc comment accordingly.
And (ironically) added tweaks from the ongoing work here (before we stop accepting them... ;).

**Comment by beorn7** [comment]
This is running on three beefy SC servers with raving success.

**Comment by beorn7** [comment]
Is there consensus that we'll go forward with this?

In that case, I'll rebase so that this is mergeable.

**Comment by beorn7** [comment]
OK, rebased.

Please let's get this done soon. This is crucial for several production servers at SoundCloud, which are unbearably slow without this improvement. I guess the same is true for others. ( @mwitkow this might also help a lot in your use case.)

Currently, I'm maintaining an internal Prometheus version just for SC, which is not the way it should be.

File: storage/local/storage.go [hunk 1]
```diff
@@ -59,6 +60,19 @@ const (
...
 	factorMinChunksToPersist = 0.2
+
+	// Threshold for when to stop using LabelMatchers to retrieve and
+	// intersect fingerprints. The rationale here is that looking up more
+	// fingerprints has diminishing returns if we already have narrowed down
+	// the possible fingerprints significantly. It is then easier to simply
+	// lookup the metrics for all the fingerprints and directly compare them
+	// to the matchers. Since a fingerprint lookup for an Equal matcher is
+	// much less expensive, there is a lower threshold for that case.
+	// TODO(beorn7): This numbers need to be tweaked, probably a bit lower.
```
Comment (juliusv) [inline storage/local/storage.go]: s/This/These/

File: storage/metric/matcher.go [hunk 2]
```diff
@@ -69,9 +78,96 @@ func NewLabelMatcher(matchType MatchType, name model.LabelName, value model.Labe
...
+	case model.InstanceLabel:
+		// Matches only metrics from a single instance, which clearly
+		// limits the damage.
+		m.score -= 0.2
+	case model.JobLabel:
+		// The usual case is a relatively low number of jobs with many
+		// metrics each.
+		m.score += 0.1
+	case model.BucketLabel, model.QuantileLabel:
+		// Magic labels for buckets and quantiles will match coupiously.
```
Comment (juliusv) [inline storage/metric/matcher.go]: coupiously -> copiously

**Comment by juliusv** [comment]
Two spelling nits was all I could find :) The rest looks indeed a bit magical, but good to me for now.

**Comment by beorn7** [comment]
Thanks for the spelling check. :->

Since this is battle-proven, I'll take @juliusv 's comment as an +1 and merge.

**Comment by grobie** [comment]
Just for the record, the impact of this change is massive! I was looking at a case where an instant query of a aggregated metric (with 4 timeseries) took 15s (possibly related to many other queries running at the same time). After upgrading to a custom version including this fix, query time went down to 100ms.

Thanks @beorn7!

**Comment by juliusv** [comment]
Woot!
